src/main/java/com/abdullah/coding/challenge/services/BookingService.java

```diff
@@ -0,0 +1,359 @@
+
+        } catch (Exception ex) {
+            return new Response<>(HttpStatus.INTERNAL_SERVER_ERROR, "Error processing the request!", null);
+        }
+    }
+
+    public Response<List<Booking>> getBookingsByCabType(String cabType) {
+        try {
+
+            List<Booking> bookings = repository.findBookingByVehicalType(cabType);
+            return new Response<>(HttpStatus.OK, "Success", bookings);
+
+        } catch (Exception ex) {
+            return new Response<>(HttpStatus.INTERNAL_SERVER_ERROR, "Error processing the request!", null);
+        }
+    }
+
+    public int deleteBookingsByStatus(int customerId, String status) {
+        return repository.deleteAllBookingsByStatus(customerId, status);
+    }
+
+    public Response<?> postComment(int bookingId, String comment) {
+        try {
+            Optional<Booking> booking = repository.findById(bookingId);
+            if (booking.isEmpty()) {
+                return new Response<>(HttpStatus.NOT_FOUND, "No booking found!", null);
+            }
+
+            Booking bookingUpdate = booking.get();
+            bookingUpdate.setComment(comment);
+            repository.save(bookingUpdate);
+
+            return new Response<>(HttpStatus.CREATED, "Comment posted!", null);
+
+        } catch (Exception ex) {
+            return new Response<>(HttpStatus.INTERNAL_SERVER_ERROR, "Error processing the request!", null);
+        }
+    }
+
+    @Transactional
+    public Response<?> postRating(int bookingId, int rating) {
+        try {
+
+            if (rating < 1 || rating > 5) {
+                return new Response<>(HttpStatus.BAD_REQUEST, "Invalid Rating", null);
+            }
+
+            Optional<Booking> booking = repository.findById(bookingId);
+            if (booking.isEmpty()) {
+                return new Response<>(HttpStatus.NOT_FOUND, "No booking found!", null);
+            }
+
+            Booking bookingUpdate = booking.get();
+
+            if (!bookingUpdate.getStatus().equals(BookingStatus.COMPLETED.toString())) {
+                return new Response<>(HttpStatus.UNPROCESSABLE_ENTITY, "Booking must be completed!", null);
+            }
+
+            if (bookingUpdate.getRating() != -1) {
+                return new Response<>(HttpStatus.UNPROCESSABLE_ENTITY, "Booking already rated!", null);
+            }
+
+            bookingUpdate.setRating(rating);
+            repository.save(bookingUpdate);
+
+            double actualRating = getCalculatedRating(bookingUpdate.getCab().getId());
+            cabService.setCabRating(actualRating, bookingUpdate.getCab().getId());
+
+            return new Response<>(HttpStatus.CREATED, "Rating posted!", null);
+
+        } catch (Exception ex) {
+            return new Response<>(HttpStatus.INTERNAL_SERVER_ERROR, "Error processing the request!", null);
+        }
+    }
+
+    private double getCalculatedRating(int cabId) {
+        List<Booking> ratedBookings = repository.getAllRatingsByCab(cabId);
```

Comment:
1. Why have we named it "getAllRatingsByCab" when we're getting the booking instead of the rating from the repository?

Reply:
It's because when the rating is calculated, the application goes through all bookings for a specific cab, fetches their rating from each booking and then calculates it by adding the total ratings of all booking of a particular cab divided by the number of times it has been booked. 

I did not introduce a new table to keep track of the ratings. Actually, it's quite similar to what I did in my sideline project but that is made in Express.js. However, the rating is much more efficient there. I can showcase the business logic if required.